python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/device_operators.cuh>
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <type_traits>
+
+template <typename T>
+__device__ bool are_all_nans(cooperative_groups::thread_block const& block,
+                             T const* data,
+                             int64_t size)
+{
+  __shared__ bool result;
+
+  if (block.thread_rank() == 0) { result = true; }
+  block.sync();
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
```

Comment:
This loop doesn't have a `#pragma unroll` but others do.

Do we know how aggressively loops like these are automatically unrolled, and/or does the `#pragma unroll` actually make a difference?

Reply:
I actually never evaluate the performance impact of `#pragma unroll` in such a scenario. Just made a quick comparison in godbolt and the SASS instructions are exactly the same. (https://godbolt.org/z/z4cjTaafT). That said, using this unroll pragma or not won't affect runtime performance IMO. The other functions use it simply because it was there when I refactor the code. Should we remove them all?

Reply:
The `#pragma unroll`  really only has an effect if the size of the loop is known at compile time.
[The CUDA Programming Guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll) mentions the expression following the pragma is expected to be an [integral constant expression](https://en.cppreference.com/w/cpp/language/constant_expression) (meaning it can be evaluated at compile-time). It also mentions the constant expression is optional but then requires the trip-count of the for-loop is constant -- known at compile-time.
Based on this, I'd recommend removing the pragma from the for-loops in this source file.

Reply:
Removed.

Reply:
Great info @davidwendt. 👍